dtree.c

```diff
@@ -661,15 +743,53 @@ void handleTextInput(SDL_Event *event){
     }
 }
 
+void moveCursorLine(int relative_line){
+    char** lines = getLines(GRAPH.selected->text.buf, 1);
+
+    int len_so_far = 0;
+    for (int i = 0; lines[i]; i++) {
+            int col = CURSOR_POSITION - len_so_far;
+            // Move Up
+            if ( relative_line == -1 && i > 0 && CURSOR_POSITION >= len_so_far && CURSOR_POSITION < len_so_far + strlen(lines[i])){
+                if ( col > strlen(lines[i-1]) ) {
+                    CURSOR_POSITION -= col + 1 + 1;
+                }
+                else {
+                    CURSOR_POSITION -= col + 1;                  // move to end of line - 1
+                    CURSOR_POSITION -= strlen(lines[i-1]) - col; //move to col'th idx of line i-1
+                }
+            }
+            // Move Down
+            else if ( relative_line == 1 && lines[i+1] && CURSOR_POSITION > len_so_far && CURSOR_POSITION < len_so_far + strlen(lines[i])){
+                if ( col > strlen(lines[i+1]) ) {
+                    CURSOR_POSITION += strlen(lines[i]) - col + strlen(lines[i+1]);
+                }
+                else {
+                    CURSOR_POSITION += strlen(lines[i]) + 1 - col;                  // move to end of line - 1
+                    CURSOR_POSITION += col; //move to col'th idx of line i-1
+                }
+                break;
+            }
+        len_so_far += strlen(lines[i]) + 1;
+    }
```

Comment:
This seems a bit verbose for just moving the cursor up or down. Maybe representing the cursor as a line number and column number would be make it easier to work with? If it must be done this way, we can still cut down on repetition by storing common checks used in each case. e.g.
`relative_line == -1 && i > 0 && CURSOR_POSITION >= len_so_far && CURSOR_POSITION < len_so_far + strlen(lines[i])`
and
`relative_line == 1 && lines[i+1] && CURSOR_POSITION > len_so_far && CURSOR_POSITION < len_so_far + strlen(lines[i])`
both use the `CURSOR_POSITION >= len_so_far && CURSOR_POSITION < len_so_far + strlen(lines[i])`
part.  (btw the second case uses `>` instead of `>=` , if this is unintentional, it is probably causing a bug)

Reply:
I looked into this a bit. I considered reworking it to be lines + cols, but that adds a lot of complexity when trying to manipulate char*s. Instead I'm going to look at just generalizing the moveCursorLine function, I had to hardcode the cases to wrap my head around it the first time around but now that I've written it before it should be easier to generalize in a refactor.

Reply:
On second thought, we don't need to generalize the moveCursorLine function, so I'm not going to right now. If I add a vim-mode for text editing, which I won't, we'll need it, but otherwise, nah. I have simplified the code though. Still need to fix bugs before pushing. New code looks like:

```
// can only handle +1 and -1
void moveCursorLine(int relative_line){
    char** lines = getLines(GRAPH.selected->text.buf, true);
    int len_so_far = 0;
    for (int i = 0; lines[i]; i++) {
            if ( CURSOR_POSITION < len_so_far || CURSOR_POSITION > len_so_far + strlen(lines[i]) ) {
                len_so_far += strlen(lines[i]) + 1;
                continue;
            }
            int col = CURSOR_POSITION - len_so_far;
            // Move Up
            if ( relative_line == -1 && i > 0 ){
                if ( col > strlen(lines[i-1]) )
                    CURSOR_POSITION -= col + 1 + 1;
                else
                    CURSOR_POSITION -= strlen(lines[i-1]) + 1;
            }
            // Move Down
            else if ( relative_line == 1 && lines[i+1] ){
                if ( col > strlen(lines[i+1]) )
                    CURSOR_POSITION += strlen(lines[i]) - col + strlen(lines[i+1]);
                else
                    CURSOR_POSITION += strlen(lines[i]) + 1;
            }
            break;
    }
    freeLines(lines);

}
```